### ISSUE TITLE
[Representation] Prototype alternatives to watch-heavy views

### ISSUE THREAD

**Issue by VWoeltjen** [nasa/openmctweb#300]
Explore alternative approaches to representing watch-heavy views (like the tree, or items view when it is dense.) Watch count is currently well over Angular's recommended limit for large trees.

Should look at some different templating systems and/or libraries here (e.g. Backbone), along with more general approaches (e.g. "observable-driven") and determine how these can integrate with Angular.

Outcome should be a variant approach which offers improved performance (at least for these kinds of views), with some guidance on when this approach is applicable and what trades should be considered. Solution should be compatible with existing view structure (with reasonable modifications)

Supports WTD-1569

**Comment by larkin** [comment]
I see a multipronged approach: 
* Use css pseudo selectors to reduce watches.
* there are some tricks with checkboxes and adjacent element selectors that can make hide/show behavior require less/no JavaScript
* use one time bindings whenever possible. 
* use observables and events to trigger state changes as opposed to scope changes. 

I think a css solution is ideal and will reduce watches significantly.

**Comment by VWoeltjen** [comment]
Some quick notes from analyzing watch count:

* Test instance contains ~700 watches during normal operation, mostly from tree (real instances contain larger trees, ~7000 watches for RP)
* 87 watches are in the bottom bar (probably similar for a full instance)
* Remainder(~500)  mostly due to browse mode
* Tree accounts for ~300 of those
* Items view for the root of that tree has ~200 watches
* Layouts have a pretty high number of watches; a layout with 2 plots had ~350 watches
* A single plot has ~90 watches

Unfortunately, these counts are mostly built up over different scopes, each of which has only a couple of watches. The tree is the worst single offender and is a good candidate to focus on, but overall this is looking like a lot of tiny fixes all over the place.

Changes to `mct-include` et al could pay back decently (for instance, `key` gets put into scope and then watched, meaning that you carry an extra watch even when `key` is a literal - refactoring to `scope.$watch(attrs.key, someKindOfUpdate)` could shave off a lot of watches)

@larkin - I've been looking out for watches that could be refactored to CSS, but haven't spotted any in regions of interest. The check box trick you're referring to is [this one](https://css-tricks.com/the-checkbox-hack/), correct? I'll try that out in the tree and see if I can build a directive about it. (The tree's a little interesting because it's not just show/hide, you want to defer loading until expansion, but think I can make that work with events.)

**Comment by VWoeltjen** [comment]
For reference, I'm digging into watches using Chrome console and some ad hoc functions:

```javascript
var root = angular.element(document.body).scope().$root;

function printCounts(scope, indent) {
   var j = indent, str = "";
   while (j--) str += " ";
   if (!scope) return;
   console.log([str, scope.$id, " ", scope.key || "?", ": ", (scope.$$watchers || []).length, " of ", scope.$$watchersCount].join(""));
   printCounts(scope.$$childHead, indent + 2);
   printCounts(scope.$$nextSibling, indent + 2);
}

findScope(scope, id) {
  if (scope && (scope.$id === id)) return scope;
  if (scope) return findScope(scope.$$nextSibling, id) || findScope(scope.$$childHead, id);
}
```

**Comment by larkin** [comment]
The check box trick is interesting for expand/collapse but still requires handling the checkbox's change event to load sub objects-- not ideal here.  Can avoid the hackyness and just use event listeners instead.

One time bindings would cut down watch counts significantly, should be liberally applied wherever a value is not assumed to change (e.g. `model.composition !== undefined`, `{{ type.getGlyph() }}`).

For heavily repeated elements:

* ng-if should always be a one-time binding.  It's only valuable to skip rendering an element if it would never be displayed, otherwise it's better to render it and toggle visibility with classes.
* ng-class should always be a one-time binding, if it would change then it would be better to do it in response to click events.
* ng-show doesn't have a purpose-- we should be using css.

So, for expand/collapse in the tree, classes should be added to parent element to reflect current state, and styles should be applied via descendent selectors to hide/show contents.

**Comment by larkin** [comment]
For ng-repeats, one time bindings inside of the repeated element are great as long as you aren't tracking by $index: For plot ticks, could use a one-time binding for the ng-style, ng-show, etc, and reduce watch counts significantly.

**Comment by VWoeltjen** [comment]
Have watch count associated with the tree down to zero on branch #303; some functionality has been lost and not restored, but this should be replaceable with the same approaches used there (observables, mainly)

**Comment by VWoeltjen** [comment]
Marking resolved; filed followup issues, linked to from #303
